platform/o.n.bootstrap/netbeans-tracking-agent/tracking-agent/src/org/netbeans/agent/TrackingAgent.java

```diff
@@ -0,0 +1,1099 @@
+                        constantPool.add(null);
+                        break;
+                    case 15:
+                        p += 3;
+                        constantPool.add(null);
+                        break;
+                    case 3: case 4: case 9: case 10: case 11:
+                    case 12: case 17: case 18:
+                        p += 4;
+                        constantPool.add(null);
+                        break;
+                    case 5: case 6:
+                        p += 8;
+                        constantPool.add(null);
+                        entry++;
+                        constantPool.add(null);
+                        break;
+                    default:
+                        System.err.println("unknown constant pool tag: " + tag);
+                        return classfileBuffer;
+                }
+            }
+
+            int cpEnd = p;
+            ByteArrayOutputStreamImpl additionalConstantPool = new ByteArrayOutputStreamImpl();
+            int[] cpLastEntry = new int[] {constantPool.size()};
+
+            p += 2; //access flags
+            p += 2; //this class
+            p += 2; //super class
+            int interfacesCount = readShort(classfileBuffer, p); p += 2;
+            p += interfacesCount * 2;
+            int fields_count = readShort(classfileBuffer, p); p += 2;
+            for (int f = 0; f < fields_count; f++) {
+                p += 2; //access flags
+                p += 2; //name
+                p += 2; //descriptor
+                p = readAttributes(constantPool, classfileBuffer, p, NOOP_INJECTOR);
+            }
+            int methods_count = readShort(classfileBuffer, p); p += 2;
+            for (int m = 0; m < methods_count; m++) {
+                p += 2; //access flags
+                int nameIdx = readShort(classfileBuffer, p); p += 2;
+                int descriptor = readShort(classfileBuffer, p); p += 2;
+                BiFunction<String, Integer, byte[]> injector = (n, pp) -> null;
+                Optional<MethodEnhancement> me = thisClassEnhancements.stream().filter(me_ -> constantPool.get(nameIdx).equals(me_.methodName)).filter(me_ -> constantPool.get(descriptor).equals(me_.methodDescriptor)).findAny();
+                if (me.isPresent()) {
+                    injector = (n, pp) -> {
+                        if (!"Code".equals(n)) {
+                            return null;
+                        }
+                        
+                        List<Integer> newConstantPoolEntries = new ArrayList<>();
+
+                        for (String cpEntry : me.get().constantPool) {
+                            newConstantPoolEntries.add(cpLastEntry[0]++);
+                            byte[] data = decodeData(cpEntry, newConstantPoolEntries);
+                            additionalConstantPool.write(data);
+                        }
+
+                        int maxStack = readShort(classfileBuffer, pp); pp += 2;
```

Comment:
On one hand, it would be easier to use ASM. On the other hand, ASM cannot handle new classfiles - we could workaround that by downgrading the versions of the classfiles sent to ASM; it would also mean to place ASM into the agent path, or doing some trampoline from the agent to a normal module. I guess I'd try with the handcoded bytecode patcher, and we'll see what are the experiences with that.